### ISSUE TITLE
Cannot create calendar with observance with pandas 2.1

### ISSUE THREAD

**Issue by MaelPic** [gerrymanoim/exchange_calendars#326]
Pandas 2.1 has changed behaviour of `DateTimeIndex.map` function (https://github.com/pandas-dev/pandas/pull/51934). This has an impact on the way exchange_calendar defines holidays using the `observance` arg. Currently, it seems creating such a calendar fails. You can try with the code below:

```
import exchange_calendars as ec
ec.get_calendar("XWAR") 
```

The problem is that observance functions implemented in exchange_calendars use a `DateTimeIndex` as arg (ex: https://github.com/gerrymanoim/exchange_calendars/blob/master/exchange_calendars/exchange_calendar_xwar.py#L36), but they should be `TimeStamps` (the elements of the `DateTimeIndex`).

See for ex this pure pandas code that used to work before 2.1, but does not work anymore:
```
from pandas.tseries.holiday import AbstractHolidayCalendar, Holiday
from pandas.tseries.offsets import CustomBusinessDay

rule = Holiday("holiday", month=1, day=1, observance=lambda dt_idx: dt_idx[dt_idx.year != 2000])
day = CustomBusinessDay(calendar=AbstractHolidayCalendar(rules=[rule]))
```

In 2.1, it should be written this way:
```
from pandas.tseries.holiday import AbstractHolidayCalendar, Holiday
from pandas.tseries.offsets import CustomBusinessDay

rule = Holiday("holiday", month=1, day=1, observance=lambda dt: dt if dt.year != 2000 else None)
day = CustomBusinessDay(calendar=AbstractHolidayCalendar(rules=[rule]))
```
See the difference in `observance` lambda definition.  
This code seems to be retro-compatible with previous versions of pandas.  
I believe all functions defined for observance need to be modified. 
Thanks!

**Comment by maread99** [comment]
Yup, verified, not working with pandas 2.1 (released 23/08/30).

Thank you for this @MaelPic, and for having tracked down the cause.

> This code seems to be retro-compatible with previous versions of pandas.

👍 Yup, a quick play suggests the required changes are backwards compatible to at least 1.5, the minimum supported pandas version.

If you, or anyone else, are desperate for a fix and are able to offer up a PR then please do. So long as it passes the tests I'll push it straight to the master.

**Comment by maread99** [comment]
This is now fixed on the master (#327). Hope to include in a release later this week.

**Comment by MaelPic** [comment]
that's great, thank you so much @maread99 !
Do you know when we could expect a new release that includes this fix?
Thanks again!

**Comment by maread99** [comment]
@MaelPic, included in [release 4.3](https://github.com/gerrymanoim/exchange_calendars/releases/tag/4.3).

Thanks again for bringing this one to our attention.
